lib/util.js

```diff
@@ -0,0 +1,170 @@
+  const obj = localRefResolve(jsonSchema, externalSchemas)
+  let toSwaggerProp
+  switch (container) {
+    case 'query':
+      toSwaggerProp = function (properyName, jsonSchemaElement) {
+        jsonSchemaElement.in = container
+        jsonSchemaElement.name = properyName
+        return jsonSchemaElement
+      }
+      break
+    case 'formData':
+      toSwaggerProp = function (properyName, jsonSchemaElement) {
+        delete jsonSchemaElement.$id
+        jsonSchemaElement.in = container
+        jsonSchemaElement.name = properyName
+
+        // https://json-schema.org/understanding-json-schema/reference/non_json_data.html#contentencoding
+        if (jsonSchemaElement.contentEncoding === 'binary') {
+          delete jsonSchemaElement.contentEncoding // Must be removed
+          jsonSchemaElement.type = 'file'
+        }
+
+        return jsonSchemaElement
+      }
+      break
+    case 'path':
+      toSwaggerProp = function (properyName, jsonSchemaElement) {
+        jsonSchemaElement.in = container
+        jsonSchemaElement.name = properyName
+        jsonSchemaElement.required = true
+        return jsonSchemaElement
+      }
+      break
+    case 'header':
+      toSwaggerProp = function (properyName, jsonSchemaElement) {
+        return {
+          in: 'header',
+          name: properyName,
+          required: jsonSchemaElement.required,
+          description: jsonSchemaElement.description,
+          type: jsonSchemaElement.type
+        }
+      }
+      break
+  }
+
+  return Object.keys(obj).reduce((acc, propKey) => {
+    acc.push(toSwaggerProp(propKey, obj[propKey]))
+    return acc
+  }, [])
+}
+
+function swagger2ParametersToOpenapi3 (jsonSchema) {
+  jsonSchema.schema = {}
+  jsonSchema.schema.type = jsonSchema.type
+  if (jsonSchema.type === 'object') {
+    jsonSchema.schema.properties = jsonSchema.properties
+  }
+  if (jsonSchema.type === 'array') {
+    jsonSchema.schema.items = jsonSchema.items
+  }
+  delete jsonSchema.type
+  delete jsonSchema.properties
+  delete jsonSchema.items
+  return jsonSchema
+}
+
+function localRefResolve (jsonSchema, externalSchemas) {
+  if (jsonSchema.type && jsonSchema.properties) {
+    // for the shorthand querystring/params/headers declaration
+    const propertiesMap = Object.keys(jsonSchema.properties).reduce((acc, h) => {
+      const required = (jsonSchema.required && jsonSchema.required.indexOf(h) >= 0) || false
+      const newProps = Object.assign({}, jsonSchema.properties[h], { required })
+      return Object.assign({}, acc, { [h]: newProps })
+    }, {})
+
+    return propertiesMap
+  }
+
+  // $ref is in the format: #/definitions/<resolved definition>/<optional fragment>
+  const localReference = jsonSchema.$ref.split('/')[2]
+  return localRefResolve(externalSchemas[localReference], externalSchemas)
+}
+
+function stripBasePathByServers (path, servers) {
+  servers = Array.isArray(servers) ? servers : []
```

Comment:
Wouldn't this lose the servers value if it's defined as a single entry?

Reply:
When I check the spec from swagger, servers can only accept array value.

Reply:
Should we be silently ignoring invalid value then?

Reply:
We need a document on this behavior.

Reply:
Maybe it's worth to have some default value. E.g. like it's done here https://github.com/SkeLLLa/fastify-oas/blob/master/lib/openapi/constructor.js#L163

Reply:
The empty array can skip the for loop of strip base path. If we specify the default value and we know it will remove the leading ```/```, then we need to add the leading ```/``` back. Also, it is the function for strip the base path only. I will remain the empty array as it can do the work faster.

Maybe we need a better URL handling here instead of using ```URL```. 
```If the servers property is not provided, or is an empty array, the default value would be a Server Object with a url value of /.```